### PR TITLE
Create Golang Release go tool

### PR DESCRIPTION
**Pull request by rcrozean**
*Issue #, if available:*

*Description of changes:*
These changes create a release object for eks go releases. The plan is to have a `func (r Release) NewRelease`, `func (r Release) UpdateMinorVersion`, and *potentially* `func (r Release) Backport` command. 

`NewRelease` would be for new minor version releases.
Example PR: https://github.com/rcrozean/eks-distro-build-tooling/pull/64 
PR was created with this tool using command: `eksGoRelease new --eksGoReleases=1.21.0 -v 4`

`Update` would be to update supported upstream versions when patch versions are released
Example PR: https://github.com/rcrozean/eks-distro-build-tooling/pull/67
PR was created with this tool using command: `eksGoRelease update --eksGoReleases=1.20.5 -v 4`
For the example PR an arbitrary patch version was selected.

`Backport` would be to attempt to patch eksGo versions not supported by upstream

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by rcrozean** [comment]
This pr was cut with this tool using command: `eksGoRelease release --eksGoReleases=1.21.0 -v 4`
https://github.com/rcrozean/eks-distro-build-tooling/pull/59/files#diff-cda23841869c4f5abcba2d67589fac3437147417a5d9e2e5e6812d8ad2b9bd58

File: tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+	goSpec                      = "golang.spec"
+	minorReleaseBranchFmt       = "eks-%s"
+	newMinorVersionCommitMsgFmt = "Init new Go Minor Version %s files."
+)
+
+func NewEksGoReleaseObject(versionString string) (*Release, error) {
+	splitVersion := strings.Split(versionString, ".")
+	major, err := strconv.Atoi(splitVersion[0])
+	if err != nil {
+		return nil, err
```
Comment (danbudris) [inline tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go:45]:
```suggestion
		return nil, fmt.Errorf("parsing major version: %v", err)
```

File: tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+	}
+
+	minor, err := strconv.Atoi(splitVersion[1])
+	if err != nil {
+		return nil, err
```
Comment (danbudris) [inline tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go:50]:
```suggestion
		return nil, return nil, fmt.Errorf("parsing minor version: %v", err)
```

File: tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+	}
+
+	patch, err := strconv.Atoi(splitVersion[2])
+	if err != nil {
+		return nil, err
```
Comment (danbudris) [inline tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go:55]:
```suggestion
		return nil, return nil, fmt.Errorf("parsing patch version: %v", err)
```

File: tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+		return err
+	}
+	if err := gClient.Add(gittagPath); err != nil {
+		logger.Error(err, "git add", "file", gittagPath)
+		return err
+	}
+
+	// Add fedora.go
+	fedoraFilePath := fmt.Sprintf(rpmSourcePathFmt, projectPath, r.GoMinorReleaseVersion(), fedora)
+	fedoraContent := []byte("temp")
```
Comment (danbudris) [inline tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go:204]: might be nice to have the content of these be links to the other implementations or some kind of other useful comment/documentation. That way someone can run this, generate the scaffolding, and then just follow on from there

File: tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+		logger.Error(err, "Adding fedora file", "path", fedoraFilePath)
+	}
+	if err := gClient.Add(fedoraFilePath); err != nil {
+		logger.Error(err, "git add", "file", fedoraFilePath)
+		return err
+	}
+
+	// Add golang-gdbinit
+	gdbinitFilePath := fmt.Sprintf(rpmSourcePathFmt, projectPath, r.GoMinorReleaseVersion(), gdbinit)
+	gdbinitContent := []byte("temp")
```
Comment (danbudris) [inline tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go:215]: same as above

File: tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+		SourceRepo:  constants.EksdBuildToolingRepoName,
+		PrRepo:      constants.EksdBuildToolingRepoName,
+		PrRepoOwner: sOwner,
+	}
+	prm := prManager.New(retrier, githubClient, prmOpts)
+
+	cprOpts := &prManager.CreatePrOpts{
+		CommitBranch:  r.EksGoReleaseFullVersion(),
+		BaseBranch:    "main",
+		AuthorName:    "rcrozean",
```
Comment (danbudris) [inline tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go:261]: we'll want to parameterize this or get it from the env

File: tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+	prUrl, err := prm.CreatePr(ctx, cprOpts)
+	if err != nil {
+		// This shouldn't be an breaking error at this point the PR is not open but the changes
+		// have been pushed and can be created manually.
+		logger.Error(err, "github client create pr failed. Create PR manually from github webclient", "create pr opts", cprOpts)
+	}
+
+	logger.V(3).Info("Release EKS Go Minor Version", "EKS Go Version", r.EksGoReleaseFullVersion(), "PR", prUrl)
+	return nil
+}
```
Comment (danbudris) [inline tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go:277]:
```suggestion
}

```

**Review by danbudris — COMMENTED** [review]
I love the direction this is going in, left some thoughts.

I'd be nice to have the `temp` values in the dummy files be meaningful comments or pointers to documentation. This could come in a follow-on PR but wanted to capture it here.

We do need to make sure we paramaterize the git user -- wether that's through a flag, and env var, or derived from the git config. It might be easiest to use an env var for this as it would easily support both the CLI use-case as well as the Prow plugin use case

File: tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+	"github.com/aws/eks-distro-build-tooling/tools/eksDistroBuildToolingOpsTools/pkg/git"
+	"github.com/aws/eks-distro-build-tooling/tools/eksDistroBuildToolingOpsTools/pkg/github"
+	"github.com/aws/eks-distro-build-tooling/tools/eksDistroBuildToolingOpsTools/pkg/prManager"
+	"github.com/aws/eks-distro-build-tooling/tools/eksDistroBuildToolingOpsTools/pkg/retrier"
+	"github.com/aws/eks-distro-build-tooling/tools/pkg/logger"
+)
+
+const (
+	githubRepoUrl               = "https://github.com/%s/eks-distro-build-tooling.git"
+	sOwner                      = "rcrozean"
```
[the remark this answers is not in the document]
Comment (rcrozean) [reply tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go:22]: Yes, the plan is to parameterize this s(ource)Owner variable and email.

File: tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go (hunk 4 above)
Comment (rcrozean) [reply tools/eksDistroBuildToolingOpsTools/pkg/eksGoRelease/eksGoRelease.go:204]: Even just explain how to get these files. Agreed that is better than "temp"

**Pull request merged**